Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs

```diff
@@ -0,0 +1,240 @@
+				string state = null;
+
+				var query = this.Request.Query;
+				var values = query.GetValues("code");
+
+				if ((values != null) && (values.Count == 1))
+				{
+					code = values[0];
+				}
+
+				values = query.GetValues("state");
+
+				if ((values != null) && (values.Count == 1))
+				{
+					state = values[0];
+				}
+
+				properties = this.Options.StateDataFormat.Unprotect(state);
+
+				if (properties == null)
+				{
+					return null;
+				}
+
+				// OAuth2 10.12 CSRF
+				if (this.ValidateCorrelationId(properties, this._logger) == false)
+				{
+					return new AuthenticationTicket(null, properties);
+				}
+
+				var tokenRequestParameters = new List<KeyValuePair<string, string>>()
+				{
+					new KeyValuePair<string, string>("client_id", this.Options.ClientId),
+					new KeyValuePair<string, string>("client_secret", this.Options.ClientSecret),
+					new KeyValuePair<string, string>("grant_type", "authorization_code"),
+					new KeyValuePair<string, string>("redirect_uri", this.GenerateRedirectUri()),
+					new KeyValuePair<string, string>("code", code),
+				};
+
+				var requestContent = new FormUrlEncodedContent(tokenRequestParameters);
+
+				var response = await this._httpClient.PostAsync(TokenEndpoint, requestContent, this.Request.CallCancelled);
+				response.EnsureSuccessStatusCode();
+
+				var oauthTokenResponse = await response.Content.ReadAsStringAsync();
+
+				var oauth2Token = JObject.Parse(oauthTokenResponse);
+				var accessToken = oauth2Token["access_token"].Value<string>();
+
+				if (string.IsNullOrWhiteSpace(accessToken) == true)
+				{
+					this._logger.WriteWarning("Access token was not found");
+					return new AuthenticationTicket(null, properties);
+				}
+
+				var graphResponse = await this._httpClient.GetAsync(GraphApiEndpoint + "?oauth_token=" + Uri.EscapeDataString(accessToken) + "&m=foursquare&v=" + DateTime.Today.ToString("yyyyyMMdd"), this.Request.CallCancelled);
+				graphResponse.EnsureSuccessStatusCode();
+
+				var accountstring = await graphResponse.Content.ReadAsStringAsync();
+				var accountInformation = JObject.Parse(accountstring);
+				var user = (JObject) accountInformation["response"]["user"];
+
+				var context = new FoursquareAuthenticatedContext(this.Context, user, accessToken);
+
+				context.Identity = new ClaimsIdentity(
+					new[]
+					{
+						new Claim(ClaimTypes.NameIdentifier, context.Id, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim(ClaimTypes.Name, context.Name, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:id", context.Id, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:name", context.Name, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:email", context.Email, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:twitter", context.Twitter, XmlSchemaString, this.Options.AuthenticationType)
```

Comment:
It is possible that the authenticated user does not have a Twitter account or public email, and so on. You need to surround **ever** claim in an `if` statement like [this](https://github.com/owin-middleware/OwinOAuthProviders/blob/master/Owin.Security.Providers/GitHub/GitHubAuthenticationHandler.cs#L98). If you do not, an exception will be thrown for users who are "missing" values.